packages/expo-constants/android/src/main/java/expo/modules/constants/ExponentInstallationId.java

```diff
@@ -0,0 +1,110 @@
+package expo.modules.constants;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Log;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.UUID;
+
+/**
+ * An installation ID provider - it solves two purposes:
+ * - in installations that have a legacy UUID persisted
+ *   in shared-across-expo-modules SharedPreferences,
+ *   migrates the UUID from there to a non-backed-up file,
+ * - provides/creates a UUID unique per an installation.
+ *
+ * Similar class exists in expoview and expo-notifications.
+ */
+public class ExponentInstallationId {
+  private static final String TAG = ExponentInstallationId.class.getSimpleName();
+
+  public static final String LEGACY_UUID_KEY = "uuid";
+  public static final String UUID_FILE_NAME = "expo_installation_uuid.txt";
+  private static final String PREFERENCES_FILE_NAME = "host.exp.exponent.SharedPreferences";
+
+  private String mUuid;
+  private Context mContext;
+  private SharedPreferences mSharedPreferences;
+
+  /* package */ ExponentInstallationId(Context context) {
+    mContext = context;
+    mSharedPreferences = context.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
+  }
+
+  public String getUUID() {
+    // If it has already been cached, return the value.
+    if (mUuid != null) {
+      return mUuid;
+    }
+
+    // Read from non-backed-up storage
+    File uuidFile = getNonBackedUpUuidFile();
+    try (FileReader fileReader = new FileReader(uuidFile);
+         BufferedReader bufferedReader = new BufferedReader(fileReader)) {
+      // Cache for future calls
+      mUuid = UUID.fromString(bufferedReader.readLine()).toString();
+    } catch (IOException | IllegalArgumentException e) {
+      // do nothing, try other sources
+    }
+
+    // We could have returned inside try clause,
+    // but putting it like this here makes it immediately
+    // visible.
```

Comment:
it's also better to put only the code that could throw inside of a try-clause so that errors from code like this don't accidentally get caught